Question 1. 
In the following answer, I am making python read a csv file that has missing values, as indicated in In[1], and then using df.isna() to prove there are missing values. 
As marked in Out[1] this dataset only contains missing values in 

In [1]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

As marked in Out[1] this dataset only contains missing values in the columns: 'id' and 'song', each missing 1 and 11 values respectively. 

Question 2.1 
Using the code provided by my Chatbot, and the dataset used in Question 1, I get the following number of columns and rows. 

In [3]:
import pandas as pd

url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'
df = pd.read_csv(url)

num_rows, num_columns = df.shape

print(f'Number of rows: {num_rows}')
print(f'Number of columns: {num_columns}')

Number of rows: 391
Number of columns: 11


So, basically what the code is telling me is that it is defining df.shape with the number of rows and columns in the dataset. The df.shape function is used to visualize the size of the file, which is also shown when the dataset is printed, just as it is in In[5]. 

In [5]:
print(df)

     row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye aye   villager-admiral   
1    

Question 2.2
Taking into account what the chatbot told me, I conclude the following: 

Observations definition: they refer to the each row in the dataset, and the rows refer to each of the different matters in the set. Each observation accounts for a particular (specific and unique) matter in the set. 
For example, the data set above contains information from villagers. And by looking at the number of rows, I can conclude that there are 391 different villagers included in this set.  

Variables definition: they constitute to the different aspects considered in the data set for each observation. That is why, variables are defined by the columns. Consequently the number of variables in a dataset is equal to the number of columns. 
In this example there are 11 varibles taken into account for each observation. Variables present details for each of the rows. 

Question 3
According to the chatobt, there is a variety of summaries I can code for a certain data set, such as the following

In [6]:
# Summary of the DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   row_n        391 non-null    int64 
 1   id           390 non-null    object
 2   name         391 non-null    object
 3   gender       391 non-null    object
 4   species      391 non-null    object
 5   birthday     391 non-null    object
 6   personality  391 non-null    object
 7   song         380 non-null    object
 8   phrase       391 non-null    object
 9   full_id      391 non-null    object
 10  url          391 non-null    object
dtypes: int64(1), object(10)
memory usage: 33.7+ KB


This type of summary provides me with the different data types of the set. In this case I know that I have data that is rather an object, which means it is non numerical, so words or phrases; or int64 which represents integers (whole numbers)

In [8]:
df.describe()

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

With the summary presented in Out[8] I get statistics from the dataset, only taking into account the numeric column data. This type of summary gives me the count: how many pieces of data there are, the average of each column and its standar deviation, as well as all of the quartiles (perecentages), and the minimum and maximum values. 
If I wanted to see all data, including categorical aspects, I would add an argument in the parenthesis of df.describe(include='all'). 

In [9]:
# Display the first 5 rows of the DataFrame
df.head()

row_n       id     name  gender    species birthday personality  \
0      2  admiral  Admiral    male       bird     1-27      cranky   
1      3  agent-s  Agent S  female   squirrel      7-2       peppy   
2      4    agnes    Agnes  female        pig     4-21        uchi   
3      6       al       Al    male    gorilla    10-18        lazy   
4      7  alfonso  Alfonso    male  alligator      6-9        lazy   

          song    phrase           full_id  \
0   Steep Hill   aye aye  villager-admiral   
1      DJ K.K.  sidekick  villager-agent-s   
2   K.K. House   snuffle    villager-agnes   
3   Steep Hill   Ayyeeee       villager-al   
4  Forest Life  it'sa me  villager-alfonso   

                                                 url  
0  https://villagerdb.com/images/villagers/thumb/...  
1  https://villagerdb.com/images/villagers/thumb/...  
2  https://villagerdb.com/images/villagers/thumb/...  
3  https://villagerdb.com/images/villagers/thumb/...  
4  https://villagerdb.com/images/villagers/thumb/...

This code is useful to visualize in a quick way how is the information sorted inmy dataset,without the need of going through all of the rows. 

In [10]:
# Example for a categorical column 'species'
df['species'].value_counts()


species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger         7
alligator     7
lion          7
bull          6
rhino         6
cow           4
octopus       3
Name: count, dtype: int64

This code helps me to see the different categories in a specific row. 

In [12]:
unique_values = df['species'].unique()

num_unique_values = df['species'].nunique()

print(f"Unique values: {unique_values}")
print(f"Number of unique values: {num_unique_values}")


Unique values: ['bird' 'squirrel' 'pig' 'gorilla' 'alligator' 'koala' 'eagle' 'anteater'
 'bull' 'mouse' 'cat' 'horse' 'hamster' 'kangaroo' 'wolf' 'penguin'
 'chicken' 'elephant' 'sheep' 'deer' 'tiger' 'cub' 'dog' 'bear' 'hippo'
 'duck' 'goat' 'ostrich' 'rabbit' 'lion' 'frog' 'monkey' 'rhino' 'octopus'
 'cow']
Number of unique values: 35


While the code above shows me the unique values of a specific column, meaning all of the different values inluded in the data set, and num_unique_values just adds all of the values found in unique_values, and tells me the number. 

Question 4

In this question I introduce a new dataset In[17], which includes both non-numerical values and missing values represented as NaN.

In [17]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [18]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [20]:
df.shape

(891, 15)

Now that I've used df.shape and df.describe() for the data set, I find some discrepancies in the code. 
df.describe() runs the statistics but not for the entire dataset, since instead of providing me with a statistical analysis for all 15 columns, I am just provided with 6 columns. This is because the other 9 columns use non-numerical values which can not be analyzed by this code. To actually see the statistics of the whole dataset I would need to type df.describe(include='all'). Although using this code, would not be veryuseful since I would not have results for all of the columns that use objects and not integers or floats. 

On the other side, I can see how the number of rows provided by df.shape (891) is the same in all of the the columns for the value 'count' in df.describe(), except for the age. This is due to the fact that there were only 741 passengers who reported their age in this dataset. By comparing the result from the number of rows in IN[20] and the ones in the count row in In[18], I can find the amount of numerical values in the dataset. As well as the amoun of non-numerical values in each column (missing values) in those columns by substracting the number of rows by the number shown in the count row. 

Question 5 

Attributes are useful to show different elements of an object in the data set. The attributes are run only by using the dot notation, and without the need to include (). Attributes are used when I want to only be able to see characteristics of certain aspects in my dataset. For instance, df.shape shows the size of the dataset, but does not execute any action. 

Methods are functions that execute actions taking as a base certain information from the dataset. Methods are able to calculate different operations, such as statistics results. They require the () even when there is no argument defined inside them. Such as df.descrbie() which will return the statistics for all the numerical values in the set. Or df.isna() which will return the number of missing values. 

Session Summary
1. Understanding Attributes and Methods:

Attributes: Variables associated with an object that represent its state or properties. They are accessed directly using dot notation (e.g., df.shape).
Methods: Functions associated with an object that represent actions or behaviors. They require parentheses to be called (e.g., df.head()).

2. Generating Dataset Summaries:

Basic Summary Using Pandas:
df.info(): Provides a concise summary of the DataFrame, including data types and non-null counts.
df.describe(): Generates summary statistics for numerical columns. Use include='all' to include categorical columns.
df.head(): Displays the first few rows of the DataFrame for a quick visual check.
df.value_counts(): Shows the distribution of unique values in a categorical column.
df['column_name'].unique(): Lists unique values in a column.
df['column_name'].nunique(): Counts the number of unique values in a 
column.

3. Handling Errors:

TypeError with df.shape(): df.shape is an attribute (not a method) and should be accessed without parentheses. Using df.shape directly returns a tuple with the number of rows and columns.

I was not able to create a link for this session, since I was not using an account with the chabot. 

Question 6 

df.describe() only gives numeric columns data 

count: how many pieces of data are there in each column 

mean: it is the average of all the data in the column  

std:the square root of the sample variance. The sample variance, called s, is obtained through a formula in which 
the most important thing to remember is that 1 is divided by n-1 because it is an estimator. 

min:the minimum vlalue in a colum

25%: the first percentile or quartile, usually called Q1

50%: the second percentile or quartile, usually called Q2, and it represents the mean value. 

75%: the third percentile or quartile, usually called Q3

max: the highest value in the column 

Question 7.1

Lets say I have a data set of customer information including various columns, such as:

- CustomerID
- Name
- Email
- PurchaseAmount
- DateOfBirth

and some of the rows have missing values in one or more columns. 

In [25]:
import pandas as pd 
url = "https://docs.google.com/spreadsheets/d/1AX0_2NcWcziTf2pnuOVaI6-Q0O6jNmH0dBPuFAu7gfM/export?format=csv"
print(url)

pd.read_csv(url)

https://docs.google.com/spreadsheets/d/1AX0_2NcWcziTf2pnuOVaI6-Q0O6jNmH0dBPuFAu7gfM/export?format=csv


CustomerID         Name                    Email  PurchaseAmount  \
0        1001  Alice Smith  alice.smith@example.com          120.50   
1        1002    Bob Brown    bob.brown@example.com           55.00   
2        1003  Carol White  carol.white@example.com             NaN   
3        1004  David Black  david.black@example.com           78.30   
4        1005   Emma Davis                      NaN          250.00   
5        1006  Frank Moore  frank.moore@example.com             NaN   
6        1007    Grace Lee    grace.lee@example.com           99.99   

  DateOfBirth  
0  1985-04-12  
1         NaN  
2  1990-08-22  
3  1982-11-15  
4  1975-06-30  
5         NaN  
6  1995-03-10

df.dropna() is used mostly to remove columns that have missing values, in this case represented with NaN

For this exercise, I want to stay with all the customers that have entered a information for each column. 
So I prefer to use df.dropna() in order to be able to look at the customers whose information would be useful for me 
to contact them, add them to a list, or basically anything in which I need the full infomration. 

In [26]:
import pandas as pd 
url = "https://docs.google.com/spreadsheets/d/1AX0_2NcWcziTf2pnuOVaI6-Q0O6jNmH0dBPuFAu7gfM/export?format=csv"
print(url)

customer_info=pd.read_csv(url)

print(customer_info.head())

https://docs.google.com/spreadsheets/d/1AX0_2NcWcziTf2pnuOVaI6-Q0O6jNmH0dBPuFAu7gfM/export?format=csv
   CustomerID         Name                    Email  PurchaseAmount  \
0        1001  Alice Smith  alice.smith@example.com           120.5   
1        1002    Bob Brown    bob.brown@example.com            55.0   
2        1003  Carol White  carol.white@example.com             NaN   
3        1004  David Black  david.black@example.com            78.3   
4        1005   Emma Davis                      NaN           250.0   

  DateOfBirth  
0  1985-04-12  
1         NaN  
2  1990-08-22  
3  1982-11-15  
4  1975-06-30  


In [27]:
customer_info.dropna()

CustomerID         Name                    Email  PurchaseAmount  \
0        1001  Alice Smith  alice.smith@example.com          120.50   
3        1004  David Black  david.black@example.com           78.30   
6        1007    Grace Lee    grace.lee@example.com           99.99   

  DateOfBirth  
0  1985-04-12  
3  1982-11-15  
6  1995-03-10

As seen in In[27], when using .dropna() I am left with all of the rows 
from the customers that filled the entire columns. While if I had used del df['col'], 
I would have erased all of the columns that had missing values, and would have created
a customer information data set with missing information from the clients. 

Question 7.2

If from the customer_info data set I just wanted to know the Name, the email and the amount of purchase of each client, then I would use del df[['CustomerID', 'DateOfBirth']], which would result in the removal of those two columns, and would let me stay with the information I want to use. 

In [30]:
del customer_info['DateOfBirth']

KeyError: 'DateOfBirth'

In [31]:
print(customer_info.columns)

Index(['CustomerID', 'Name', 'Email', 'PurchaseAmount'], dtype='object')


Question 7.3 and 7.4

When using both df.dropna() and del df['col'], it is important to start with 
del df['col']since that will allow me to erase first all of the unnecesary 
columns in the data set. And once I have done that, then I can erase all of 
the rows that have missing values from the remaining columns. That way, I will
be left with the necessary columns with the complete information in each of them. 

In [ ]:
ChatBot (ChatGPT) Summary until this point 
#Summary
In this session, we explored how to load a CSV file from a Google Sheets URL into a pandas DataFrame in a Jupyter Notebook. We provided a sample dataset with missing values and demonstrated how to handle it.


#Dataset Example
We created a sample dataset in CSV format with missing values in some columns. The data included columns like CustomerID, Name, Email, PurchaseAmount, and DateOfBirth.

#Code to Load CSV from Google Sheets
Below is the code snippet to load a CSV file from a Google Sheets URL into a pandas DataFrame named customer_info. Ensure you have pandas installed in your environment.

python
Copy code
import pandas as pd

# Define the URL of the CSV file
url = 'https://docs.google.com/spreadsheets/d/1AX0_2NcWcziTf2pnuOVaI6-Q0O6jNmH0dBPuFAu7gfM/export?format=csv'

# Load the CSV file into a DataFrame
customer_info = pd.read_csv(url)

# Display the first few rows of the DataFrame to verify it's loaded correctly
print(customer_info.head())
Instructions
Import pandas: Ensure pandas is imported at the beginning of your notebook.
Load the CSV: Use pd.read_csv(url) to load the CSV file from the provided URL into a DataFrame.
Verify the DataFrame: Use print(customer_info.head()) to inspect the first few rows and confirm that the data is loaded correctly.
Feel free to modify and run the provided code in your Jupyter Notebook to load and explore the dataset.

This summary includes the relevant information and code to load the CSV file from Google Sheets into a pandas DataFrame, which you can directly copy and use in your Jupyter Notebook.

Question 8

The following data set is the one provided as an example by the ChatBot. 

In [32]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group by 'Pclass' and describe the 'Age' column within each group
age_summary_by_class = df.groupby('pclass')['age'].describe()

# Display the result
print(age_summary_by_class)


        count       mean        std   min   25%   50%   75%   max
pclass                                                           
1       186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
2       173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
3       355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0


In [33]:
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [34]:
print(df.columns)

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')


I used print(df.columns) in order to see the way in which 'pclass' is written 
in the code in order to be able to run In[32]

Question 8.1

Now I will use my own example from the titanic set to demonstrate what  df.groupby("col1")["col2"].describe() does 

In [35]:
df.groupby('fare')['survived'].describe()

count      mean       std  min   25%  50%   75%  max
fare                                                          
0.0000     15.0  0.066667  0.258199  0.0  0.00  0.0  0.00  1.0
4.0125      1.0  0.000000       NaN  0.0  0.00  0.0  0.00  0.0
5.0000      1.0  0.000000       NaN  0.0  0.00  0.0  0.00  0.0
6.2375      1.0  0.000000       NaN  0.0  0.00  0.0  0.00  0.0
6.4375      1.0  0.000000       NaN  0.0  0.00  0.0  0.00  0.0
...         ...       ...       ...  ...   ...  ...   ...  ...
227.5250    4.0  0.750000  0.500000  0.0  0.75  1.0  1.00  1.0
247.5208    2.0  0.500000  0.707107  0.0  0.25  0.5  0.75  1.0
262.3750    2.0  1.000000  0.000000  1.0  1.00  1.0  1.00  1.0
263.0000    4.0  0.500000  0.577350  0.0  0.00  0.5  1.00  1.0
512.3292    3.0  1.000000  0.000000  1.0  1.00  1.0  1.00  1.0

[248 rows x 8 columns]

In this example, df.groupby('fare') is referring to the groups that the 
column survived had, which range from the 0.0 all the way to the 512.3293. 
While ['survived'].describe() is taking into account the survived column 
for each of the different values of fare in the dataset. 

So basically what df.groupby('fare')['survived'].describe() is doing is to 
calculate the statistics (count, mean, std, min, 25%,50%,75%, and max) of
the passangers that survived, depending on the fare they paid. 
For example, from the people that paid 227.5250 for a fare,the
mean of passengers that survived is 0.75. 

In [ ]:
Summary until this point 
### Session Summary:

In this session, you aimed to analyze the Titanic dataset using pandas, focusing on understanding the functionality of the `df.groupby("col1")["col2"].describe()` code.

#### Key Points Covered:

1. **Groupby and Describe:**
   - We discussed how `df.groupby("Pclass")["Age"].describe()` works, explaining that it groups the dataset by passenger class (`Pclass`) and then generates summary statistics for the `Age` column within each class.

2. **Troubleshooting a KeyError:**
   - You encountered a `KeyError: 'Pclass'`, which we identified as potentially being caused by the column name not matching what you expected. You were advised to:
     - Check the DataFrame's column names using `df.columns`.
     - Strip any extra spaces from the column names if necessary.
     - Reload the dataset to ensure it contains the correct columns.

3. **Coding Results:**
   - Although there was an error, we reviewed steps to troubleshoot and resolve it, including how to inspect and correct column names in the DataFrame.

### Access to Code:

You can access and run the code via this Google Colab link: [Titanic Dataset Analysis](https://colab.research.google.com/). Just copy the code we discussed and paste it into a new Colab notebook to continue your analysis.


Question 8.2 

The main diference encountered when using df.groupby("col1")["col2"].describe(), instead of just df.describe is the following: 

When using df.describe the count value takes into account all of the values (numerical) across all of the rows from the dataset. While when using the df.groupby, when running the code, the count value will just analyze how the rows are presented in col1, and apply that to col2. 

Essentially, df.describe() shows an entire view of the missing values in each column. While df.groupby will only show how col2 non-missing data is distributed among the different categories of col1. 

Question 8.3A

Forgetting to include import pandas as pd

By using Chatgpt, I was able to immediately locate what ModuleNotFoundError: No module named 'pd' meant
which basically refers to the fact that i forgot to write import pandas as pd in Int[36] line 1. 

It took just a few seconds to the chatbot to solve my error. 

Nonetheless, when I looked for the meaning of the error in google, it also provided me in a matter of 
seconds the reasoning behind the error. Although it did not give a written code solution like the chatbot,
it did showed in a clear and hihglithed sentence what i was missing, by stating 
Python cannot locate the Pandas library. 

In [36]:
import pd 

url="https://github.com/mwaskom/seaborn-data/blob/master/penguins.csv"
pd.read_csv(url)

ModuleNotFoundError: No module named 'pd'

Question 8.3B

Different typo errors in the url

For all of the typos in the url, it always shows the same error: [Errno 2] No such file or directory: 'TitanicS.csv' 
which means that there is no file with the used name in the memory. 
When trying to troubleshoot this error with the chatbot, it tells me to check the current file path, verify the name, 
check the location of th file and the permissions. It provides me with a series of different aspects from my notebook 
that I can check and verify in order to solve the error. 
However, in this matter, the search from google gave more complete solutions, since in the second website I visited 
it provided the same list of aspects to check, but each aspect with an example of code typed in Python3, which results
extremely useful for me since it is the program that I am able to understand. 

Different examples of typo mistakes used fromm In[38-40]

In [38]:
import pandas as pd
url = 'titanics.csv'
pd.read_csv(url)

FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'

In [39]:
import pandas as pd
url = 'titanic.scv'
pd.read_csv(url)

FileNotFoundError: [Errno 2] No such file or directory: 'titanic.scv'

In [40]:
import pandas as pd
url = 'TitanicS.csv'
pd.read_csv(url)

FileNotFoundError: [Errno 2] No such file or directory: 'TitanicS.csv'

Question 8.3C

For this problem, the chatbot gave me the quick and easy solution to understand
since it showed me how I should define DF for the code to work. While google
sent me to a website in whcih different people published their responses and 
solutions and it was harder to understand, and also longer. 

In [42]:
DF.groupby('col1')['col2'].describe()

NameError: name 'DF' is not defined

Question 8.3D

Again in this case, the chatbot told me where my mistake was, and showed me 
the correct and complete way to write my cell code. In the google search I 
was not able to locate my answer, and I had to spend more time looking for a 
conscice answer. 

In [43]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url

SyntaxError: incomplete input (1974001384.py, line 5)

Question 8.3E

For this AttributeError, the chatbot gave the quiker answer with just 
enough infomraiton to understand where the mistake was and solve it. In 
fact it told how describe had a typo error with describle. 

In [45]:
df.groupby('fare')['survived'].describle()

AttributeError: 'SeriesGroupBy' object has no attribute 'describle'

Question 8.3F

For this KeyError, the chatbot was again more helpful than the self search. 
First it told me to write the code in In[47], and after seeing the names of 
all the columns my dataset had, then it told me the correct writing for the 
code, being df.groupby('fare')['correct_column_name'].describe(), which solves
the issue. 

In [46]:
df.groupby('fare')['name'].describe()

KeyError: 'Column not found: name'

In [47]:
print(df.columns)

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')


Question 8.3G

With this NameError, the chatbot was still useful, since it provided the
information that a column name is not a standalone variable, which is how 
it is coded in In[48], and it provided with the correct code. Google was 
also useful since it provided me with websites a short vides on how to solve
the problem.

In [48]:
df.groupby(fare)['age'].describe()

NameError: name 'fare' is not defined

Conclusion on Question8.3

Looking at all the results from A-G, my go to resource to solve errors 
is the chabot, ChatGPT, since it provides me with short and clear answers
in just a few seconds, helping me to correct the error in my code fast. 
While Google is also a useful resource, it takes more time and shows more 
complex solutions. 

In [ ]:
Summary for quesion 8
Here's a summary of our session:

 **NameError**:
   - You encountered a `NameError: name 'pd' is not defined`, which was resolved by ensuring that `pandas` was imported with `import pandas as pd`.

 **FileNotFoundError**:
   - You had a `FileNotFoundError: [Errno 2] No such file or directory: 'TitanicS.csv'`. This was addressed by checking the file path, ensuring the file exists, and verifying the file name.

 **SyntaxError**:
   - The `SyntaxError: incomplete input` error was due to incomplete or incorrect syntax. The solution involved checking for missing parentheses, brackets, and ensuring complete statements.

 **AttributeError**:
   - You encountered `AttributeError: 'SeriesGroupBy' object has no attribute 'describle'`. This was caused by a typo; the correct method is `describe()`.

 **KeyError**:
   - The `KeyError: 'Column not found: name'` was due to an incorrect column name. It was resolved by verifying the actual column names in the DataFrame and using the correct name.

 **NameError for Column Name**:
   - The `NameError: name 'fare' is not defined` occurred because `'fare'` was used as if it were a variable. The solution involved ensuring `'fare'` is a column name in the DataFrame and using it correctly within DataFrame methods.

If you need more details or have additional questions, feel free to ask!